File: src/main/java/com/procurement/access/model/dto/ocds/ProcurementMethod.java
package com.procurement.access.model.dto.ocds;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.procurement.access.exception.EnumException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public enum ProcurementMethod {
    OPEN("open"),
    SELECTIVE("selective"),
    LIMITED("limited"),
    DIRECT("direct");

    private static final Map<String, ProcurementMethod> CONSTANTS = new HashMap<>();
    private final String value;

    static {
        for (final ProcurementMethod c : values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    ProcurementMethod(final String value) {
        this.value = value;
    }

    @JsonCreator
    public static ProcurementMethod fromValue(final String value) {
        final ProcurementMethod constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new EnumException(ProcurementMethod.class.getName(), value, Arrays.toString(values()));
        }
        return constant;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @Override
    public String toString() {
        return this.value;
    }
}

File: src/main/kotlin/com/procurement/access/domain/model/enums/ProcurementMethod.kt
package com.procurement.access.domain.model.enums

import com.fasterxml.jackson.annotation.JsonValue
import com.procurement.access.domain.fail.error.DataErrors
import com.procurement.access.domain.util.Result
import com.procurement.access.domain.util.asFailure
import com.procurement.access.domain.util.asSuccess
import com.procurement.access.exception.EnumElementProviderException

enum class ProcurementMethod(@JsonValue val key: String) {
    CD("selective"),
    CF("selective"),
    DA("limited"),
    DC("selective"),
    FA("limited"),
    GPA("selective"),
    IP("selective"),
    MV("open"),
    NP("limited"),
    OF("selective"),
    OP("selective"),
    OT("open"),
    RT("selective"),
    SV("open"),
    TEST_CD("selective"),
    TEST_CF("selective"),
    TEST_DA("limited"),
    TEST_DC("selective"),
    TEST_FA("limited"),
    TEST_GPA("selective"),
    TEST_IP("selective"),
    TEST_MV("open"),
    TEST_NP("limited"),
    TEST_OF("selective"),
    TEST_OP("selective"),
    TEST_OT("open"),
    TEST_RT("selective"),
    TEST_SV("open");

    override fun toString(): String = key

    companion object {

        private val allowedValues = values()

        fun creator(name: String) = try {
            valueOf(name)
        } catch (ignored: Exception) {
            throw EnumElementProviderException(
                enumType = this::class.java.canonicalName,
                value = name,
                values = allowedValues.joinToString { it.name }
            )
        }

        fun tryCreate(name: String): Result<ProcurementMethod, DataErrors> = try {
            valueOf(name)
                .asSuccess()
        } catch (ignored: Exception) {
            DataErrors.Validation.UnknownValue(
                name = "pmd",
                expectedValues = allowedValues.map { it.name },
                actualValue = name
            )
                .asFailure()
        }
    }
}
